### PR TITLE
fix(cookies-ios): always set cookies as request header

### PR DESCRIPTION
**Pull request by stefan-sabau-3pg**
Post iOS 11, the `WebKit` API introduces specific `setCookie:completion:` methods to the `WKWebSiteDataStore` for setting the session cookies of a `WKWebView`
A common approach is to save the cookies in the `NSHTTPCookieStorage`'s `shared instance
Then using the above method set them in the appropriate `httpCookieStore` in the `websiteDataStore` of the `WKWebViewConfig` that is used to initialise the Webview
Although the cookies eventually get set in the store, those cookies do not reliably get received by the `WKWebView`
Even if trying to add specific locking/synchronisation code to make sure all `setCookie:completion:` calls are done before the WebView init, the first load of the WebView does not have the cookies, only subsequent loads will contain them
The issue is present when using the default `WKWebsiteDataStore`
The `nonPersistent` data store is not affected by this, but using this  could potentially cause other issues as this means that all data saved by the webpage in `localStorage` or `sessionStorage` could be lost when navigating or refreshing

Because cookies are no more than a special formatted request header, the equivalent of the above can be achieved by simply adding them in the header of the desired `NSURLRequest`s that the WebView handles
This is also the the most common approach pre iOS 11, before having specific methods for setting cookies in WebKit's APIs
Similarly to the above, before creating the WebView, the `NSHTTPCookieStorage` shared instance is used to get all the saved cookies
Then use each cookie to build a JS function that generates the request header and use this as a `WKUserScript` that is set to run at document start and for all frames

https://bugs.webkit.org/show_bug.cgi?id=185624
https://bugs.webkit.org/show_bug.cgi?id=185483
https://forums.developer.apple.com/message/346511#346511
https://forums.developer.apple.com/message/346450#346450

### CODE REVIEW

File: ios/RNCWebView.m [hunk 1]
```diff
@@ -238,6 +225,9 @@ - (void)didMoveToWindow
           [script appendFormat:@"document.cookie = %@ + '=' + %@",
             RCTJSONStringify(cookie.name, NULL),
             RCTJSONStringify(cookie.value, NULL)];
+          if (cookie.domain) {
+            [script appendFormat:@" + '; Domain=' + %@", RCTJSONStringify(cookie.domain, NULL)];
```
Comment (maxiconil) [inline ios/RNCWebView.m:235]: Shouldn't we check that the cookie domian is the same than the request one?

**Review by comanzanares — APPROVED** [review]
Looking good! great job

**Review by eelmoni — APPROVED** [review]

**Review by maxiconil — APPROVED** [review]
LGTM!

File: ios/RNCWebView.m (hunk 1 above)
Comment (stefan-sabau-3pg) [reply ios/RNCWebView.m:235]: Good question.
I've added this snippet when retrieving cookies, before even building the script:

```
NSURLRequest *request = [RCTConvert NSURLRequest:self.source];

NSArray<NSHTTPCookie *> *cookies = (request != nil && request.URL != nil) ?
            [[NSHTTPCookieStorage sharedHTTPCookieStorage] cookiesForURL:request.URL]
            : [[NSHTTPCookieStorage sharedHTTPCookieStorage] cookies];
```

Which is pretty much like the handling from the `requestForSource:` method: If the `WebView` needs to load an URL, convert the JSON passed from RN as params to a `NSURLRequest` object then check its domain.

**Review by maxiconil — APPROVED** [review]
Go ahead!

**Pull request merged**